src/imitation/algorithms/preference_comparisons.py

```diff
@@ -410,13 +405,26 @@ def __init__(
             reward_trainer: trains the reward model based on pairs of fragments and
                 associated preferences. Default is to use the preference model
                 and loss function from DRLHP.
+            custom_logger: Where to log to; if None (default), creates a new logger.
         """
+        self.logger = custom_logger or logger.configure()
         self.model = reward_model
-        self.reward_trainer = reward_trainer or CrossEntropyRewardTrainer(reward_model)
+        self.reward_trainer = reward_trainer or CrossEntropyRewardTrainer(
+            reward_model, custom_logger=self.logger
+        )
+        # If the reward trainer was created in the previous line, we've already passed
+        # the correct logger. But if the user created a RewardTrainer themselves and
+        # didn't manually set a logger, it would be annoying if a separate one was used.
+        if hasattr(self.reward_trainer, "logger"):
```

Comment:
You're right, I think this is bad even in Python (for exactly the reason you describe). I went with 1. now because most subclasses seem to use loggers anyway. Had to turn `Fragmenter` etc., which where previously just type aliases into actual ABCs but maybe that would have made sense anyway at some point.